datalad/interface/rerun.py

```diff
@@ -190,54 +190,67 @@ def __call__(
                 message="cannot rerun history with merge commits")
             return
 
-        Revision = namedtuple("Revision", ["id", "message", "info"])
-
-        def revision_with_info(rev):
-            msg, info = get_commit_runinfo(ds.repo, rev)
-            return Revision(rev, msg, info)
-
-        ids = ds.repo.repo.git.rev_list("--reverse", revrange, "--").split()
-
-        try:
-            revs = list(map(revision_with_info, ids))
-        except ValueError as exc:
-            yield dict(err_info, status='error', message=exc_str(exc))
-            return
+        revs = [{"hexsha": hexsha,
+                 "message": ds.repo.repo.git.show(
+                     hexsha, "--format=%B", "--no-patch")}
+                for hexsha in ds.repo.repo.git.rev_list(
+                        "--reverse", revrange, "--").split()]
+
+        for rev in revs:
+            try:
+                subj, info = get_run_info(rev["message"])
+            except ValueError as exc:
+                yield dict(err_info, status='error',
+                           message="Error on {}'s message: {}".format(
+                               rev["hexsha"], exc_str(exc)))
+                return
+            if info is not None:
+                rev["run_info"] = info
+                rev["run_subject"] = subj
 
         if since is not None and since.strip() == "":
             # For --since='', drop any leading commits that don't have
             # a run command.
-            revs = list(dropwhile(lambda r: r.info is None, revs))
+            revs = list(dropwhile(lambda r: "run_info" not in r, revs))
 
         if script:
-            commands = (r.info["cmd"] for r in revs if r.info is not None)
-            with open(script, 'w') as ofh:
-                header = """\
+            ofh = sys.stdout if script.strip() == "-" else open(script, "w")
+            header = """\
 #!/bin/sh
 #
 # This file was generated by running (the equivalent of)
 #
 #   datalad rerun --script={script}{since} {revision}
 #
-# in {path}\n\n"""
-                ofh.write(header.format(
-                    script=script,
-                    since="" if since is None else " --since=" + since,
-                    revision=ds.repo.repo.git.rev_parse(revision),
-                    path=ds.path))
-
-                for cmd in commands:
-                    if isinstance(cmd, list):
-                        cmd = " ".join(cmd)
-                    ofh.write(cmd + "\n")
+# in {path}\n"""
+            ofh.write(header.format(
+                script=script,
+                since="" if since is None else " --since=" + since,
+                revision=ds.repo.repo.git.rev_parse(revision),
+                path=ds.path))
+
+            for rev in revs:
+                if "run_info" not in rev:
+                    continue
+
+                ofh.write(
+                    "\n" + "".join("# " + ln
+                                   for ln in rev["message"].splitlines(True)))
```

Comment:
It would make sense to me to cut the message prior to `\n\n=== Do not change lines below ===`. The dump of the run record should be a straight duplicate of the following segment of shell code.